cpg_infra/billing_aggregator/driver.py

```diff
@@ -151,6 +152,45 @@ def create_source_archive(
             opts=pulumi.ResourceOptions(replace_on_changes=['*']),
         )
 
+    def setup_gcp_cost_control(self):
+        """
+        Create the gcp cost control cloud function to cut off billing when
+        it exceeds the budget
+        """
+        region = self.config.gcp.region
+        service_account = self.config.billing.gcp_cost_controls.machine_account
+        slack_channel = self.config.billing.gcp_cost_controls.slack_channel
+        pubsub_topic_name = self.config.billing.gcp_cost_controls.pubsub_topic
+
+        # Create source archive
+        source_archive = self.create_source_archive(
+            'billing-gcp-cost-report-source-code',
+            self.source_bucket.name,
+            os.path.join(os.path.dirname(__file__), 'gcp_cost_control'),
+        )
+
+        # Deploy Cloud Function
+        function = gcp.cloudfunctionsv2.Function(
+            'gcp-cost-control',
+            service_account_email=service_account,
+            source_archive_bucket=self.source_bucket.name,
+            source_archive_object=source_archive,
+            source_file='main.py',
+            entry_point='gcp_cost_control',
+            runtime='python311',
+            available_memory_mb=256,
+            environment_variables={
+                'SLACK_CHANNEL': slack_channel,
+            },
+            event_trigger=gcp.cloudfunctionsv2.FunctionEventTriggerArgs(
+                event_type='google.pubsub.topic.publish',
+                pubsub_topic=pubsub_topic_name,
```

Comment:
I _think_ it might be ok, the docs for the pulumi config aren't very clear but the cli docs say that it doesn't need the full name including the project:
https://cloud.google.com/sdk/gcloud/reference/functions/deploy#--trigger-topic

Reply:
Yup the pubsub_topic takes the topic name only. I believe because it needs to be in the same project/region anyway.